exonum/tests/map_index_prop.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// cspell:ignore oneof
+
+//! Property testing for map index and proof map index as a rust collection.
+
+extern crate exonum;
+#[macro_use]
+extern crate proptest;
+
+use exonum::storage::{Database, Fork, MapIndex, MemoryDB};
+use proptest::{collection::vec, num, prelude::*, strategy};
+
+use std::collections::HashMap;
+
+#[derive(Debug, Clone)]
+enum Action {
+    //Applied to key modulo 8
+    Put(u8, i32),
+    //Applied to key modulo 8
+    Remove(u8),
+    Clear,
+    MergeFork,
+}
+
+impl Action {
+    fn apply_map(&self, map: &mut MapIndex<&mut Fork, u8, i32>, ref_map: &mut HashMap<u8, i32>) {
+        match *self {
+            Action::Put(k, v) => {
+                let k = k % 8;
+                map.put(&k, v);
+                ref_map.insert(k, v);
+            }
+            Action::Remove(k) => {
+                let ref k = k % 8;
+                map.remove(k);
+                ref_map.remove(k);
+            }
+            Action::Clear => {
+                map.clear();
+                ref_map.clear();
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+proptest!{
+    #[test]
+    fn proptest_map_index_to_rust_map(ref actions in
+                     vec( prop_oneof![
+                         (num::u8::ANY, num::i32::ANY).prop_map(|(i, v)| Action::Put(i,v)),
+                         num::u8::ANY.prop_map(Action::Remove),
+                         strategy::Just(Action::Clear),
+                         strategy::Just(Action::MergeFork),
+                     ] , 1..10) ) {
+        let db = MemoryDB::new();
+
+        let mut fork = db.fork();
+        let mut ref_map : HashMap<u8,i32> = HashMap::new();
+
+        for action in actions {
```

Comment:
I thought about it but it looks a little bit redundant since proptest after generating vector of actions and failing will reduce it by popping actions from the back of the vector -> we will see the same output as if we would compare after each action.

Reply:
It seems to me that early fail in this case may be more "intuitive", so to speak. It may be the case (however rare) that the list of actions as a whole leads to identical results, but a prefix of this list isn't.